### ISSUE TITLE
Consider training parameters by optimising reward function

### ISSUE THREAD

**Issue by RobinL** [RobinL/uk_address_matcher#36]

**Comment by RobinL** [comment]
import duckdb
import numpy as np

# Connect to DuckDB (assuming a reward function exists)
con = duckdb.connect(database=":memory:")

def black_box_reward(params):
    """
    Computes the reward function in DuckDB.
    This is a placeholder; replace it with your actual SQL computation.
    """
    params_str = ', '.join(map(str, params))
    
    query = f"SELECT some_reward_function({params_str}) AS reward"
    result = con.execute(query).fetchone()
    
    return result[0] if result else float('-inf')  # Handle query failure

# SPSA Hyperparameters
alpha = 0.1  # Learning rate
epsilon = 0.01  # Initial perturbation magnitude
momentum = 0.9  # Momentum factor
num_iterations = 100  # Number of iterations
num_params = 50  # Number of parameters

# Initialize parameters, velocity, and adaptive perturbation scaling
params = np.random.uniform(-1, 1, num_params)
velocity = np.zeros(num_params)  # Momentum term
perturb_scale = np.ones(num_params) * epsilon  # Adaptive perturbation scale

for iteration in range(num_iterations):
    # Generate an adaptive perturbation vector
    delta = np.random.choice([-1, 1], size=num_params) * perturb_scale

    # Compute function values at perturbed points
    params_plus = params + delta
    params_minus = params - delta

    reward_plus = black_box_reward(params_plus)
    reward_minus = black_box_reward(params_minus)

    # Estimate gradient with SPSA
    gradient = (reward_plus - reward_minus) / (2 * delta)

    # Apply momentum: accumulate velocity
    velocity = momentum * velocity + alpha * gradient

    # Update parameters
    params -= velocity

    # Adaptive perturbation scaling: Increase for parameters with high gradient variance
    perturb_scale = 0.9 * perturb_scale + 0.1 * np.abs(gradient)

    # Logging progress
    if iteration % 10 == 0:
        print(f"Iteration {iteration}: Reward = {reward_plus}, Params Sample = {params[:5]}")

# Final optimized parameters
print("Optimized Parameters:", params)

**Comment by RobinL** [comment]
Working on this here:
https://github.com/RobinL/uk_address_matcher/tree/gradient_descent
